pkg/kubelet/kuberuntime/logs/filter_log.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logs
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// circularLinkedList is an append-only linked list, and behaves like a ring buffer.
+type circularLinkedList struct {
```

Comment:
A new data structure needs a lot of testing.  I'd expect significant unit tests for the data structure verifying the contract.

Also, can you more clearly justify why you need a new data structure vs reusing one of the existing stdlib data structures in comments inline?  I'm hazy on why `container/ring` is not sufficient - a comment inline justifying it would be preferred.

Reply:
Actually I have considered `container/ring` before, I decided to add this new data structure mainly for memory efficiency.

`container/ring` allocates all nodes on creation, this might not be sufficient. As for the current implementation, we would use the `tailLines` specified by user as the capacity of the circular list. Let's say a log file only contains 10 lines, and the `tailLines` is set to 1000 by user. `container/ring` would allocate 1000 elements on creation which is not quite memory efficient, while the added `circularLinkedList` would only create new element if necessary.

Reply:
How many tail logs would we have to have concurrently before that is an issue?   I would suspect the per connection request overhead for those would be significantly higher than container/ring, and so if the only purpose of this data structure is to lazily allocate elements, that's hard to justify against the extra complexity here.  Do you have numbers to support why the efficiency of the ring buffer here is the key issue?